Add Alcatel-Lucent Enterprise OmniSwitch AOS support

gmcms opened the pull request:
Adding support for Alcatel-Lucent Enterprise OmniSwitch

ktbyers commented:
@gmcms Were you able to run the unit tests running?

It would be good to verify these two sets of unit tests:
```
py.test -v test_netmiko_show.py --test_device aos_device
py.test -v test_netmiko_config.py --test_device aos_device
```

Can you please post the output of the unit tests here?

gmcms commented:
================================================================= test
session starts
=================================================================
platform linux2 -- Python 2.7.12+, pytest-2.9.2, py-1.4.31, pluggy-0.3.1 --
/usr/bin/python
cachedir: ../.cache
rootdir: /home/lubuntu/Bureau/netmiko, inifile:
collected 12 items

test_netmiko_show.py::test_disable_paging PASSED
test_netmiko_show.py::test_ssh_connect PASSED
test_netmiko_show.py::test_ssh_connect_cm PASSED
test_netmiko_show.py::test_send_command_timing PASSED
test_netmiko_show.py::test_send_command_expect PASSED
test_netmiko_show.py::test_base_prompt PASSED
test_netmiko_show.py::test_strip_prompt PASSED
test_netmiko_show.py::test_strip_command PASSED
test_netmiko_show.py::test_normalize_linefeeds PASSED
test_netmiko_show.py::test_clear_buffer PASSED
test_netmiko_show.py::test_enable_mode PASSED
test_netmiko_show.py::test_disconnect PASSED

============================================================= 12 passed in
33.51 seconds ==============================================================

================================================================= test
session starts
=================================================================
platform linux2 -- Python 2.7.12+, pytest-2.9.2, py-1.4.31, pluggy-0.3.1 --
/usr/bin/python
cachedir: ../.cache
rootdir: /home/lubuntu/Bureau/netmiko, inifile:
collected 7 items

test_netmiko_config.py::test_ssh_connect PASSED
test_netmiko_config.py::test_enable_mode PASSED
test_netmiko_config.py::test_config_mode PASSED
test_netmiko_config.py::test_exit_config_mode PASSED
test_netmiko_config.py::test_command_set PASSED
test_netmiko_config.py::test_commands_from_file PASSED
test_netmiko_config.py::test_disconnect PASSED

============================================================== 7 passed in
15.12 seconds ==============================================================


2017-06-04 19:11 GMT+02:00 Kirk Byers <notifications@github.com>:

> @gmcms <https://github.com/gmcms> Were you able to run the unit tests
> running?
>
> It would be good to verify these two sets of unit tests:
>
> py.test -v test_netmiko_show.py --test_device aos_device
> py.test -v test_netmiko_config.py --test_device aos_device
>
> Can you please post the output of the unit tests here?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ktbyers/netmiko/pull/486#issuecomment-306053252>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AGiMeQhlRTn3u6HG1u91w3PwbSc58GoEks5sAuVJgaJpZM4NvbP3>
> .
>

ktbyers commented on netmiko/alcatel/alcatel_aos_ssh.py:
```diff
@@ -0,0 +1,42 @@
+    def session_preparation(self):
+        # Prompt can be anything, but best practice is to end with > or #
+        self._test_channel_read(pattern=r'[>#]')
+        self.set_base_prompt()
```
Does AOS do any paging of output...for example prompt for `more`?

ktbyers commented on netmiko/alcatel/alcatel_aos_ssh.py:
```diff
@@ -0,0 +1,42 @@
+    """Alcatel-Lucent Enterprise AOS support."""
+    def __init__(self, *args, **kwargs):
+        super(AlcatelAosSSH, self).__init__(*args, **kwargs)
+        self._config_mode = False
```
General pattern in Netmiko is not to emulate config mode (if it doesn't exist). So I think we should get rid of the `self._config_mode` variable and not track this.

ktbyers commented:
@gmcms FYI, I removed the tracking of the config mode (since AOS doesn't support a config mode). Let me know if you see any problems with this.

See changes here https://github.com/ktbyers/netmiko/commit/ab3458a64062038a41bb9d32e79fe3953684872f

From your earlier statements, I assumed this supported AOS6 and AOS8. Let me know if this is not correct.

gmcms commented:
Hi Kirk, sorry I was not able to answer to you. I was in a DataCenter far
from home without any access to my own environment.

"Also can you add a note to the alcatel_aos_ssh.py file pertaining to which
versions of AOS you tested with.
From your earlier statements, I assumed this supported AOS6 and AOS8. Let
me know if this is not correct."

Yes it's ok for the AOS6 and AOS8. AOS7 is the same train as AOS8 which is
the new release.


"Does AOS do any paging of output...for example prompt for more?"

No, the default is not to use paging. If you want to use it, you have a
command syntax which is not the same on AOS6 and AOS8.


"General pattern in Netmiko is not to emulate config mode (if it doesn't
exist). So I think we should get rid of the self._config_mode variable and
not track this.
FYI, I removed the tracking of the config mode (since AOS doesn't support a
config mode). Let me know if you see any problems with this."

Sorry I thought that it was the right way to do this. I'm ok with your
commit.

If I can do more tests to add features or if you want to look at a way to
add it in NAPALM, you can contact me.


Thanks


2017-06-09 22:07 GMT+02:00 Kirk Byers <notifications@github.com>:

> @gmcms <https://github.com/gmcms> FYI, I removed the tracking of the
> config mode (since AOS doesn't support a config mode). Let me know if you
> see any problems with this.
>
> See changes here ab3458a
> <https://github.com/ktbyers/netmiko/commit/ab3458a64062038a41bb9d32e79fe3953684872f>
>
> From your earlier statements, I assumed this supported AOS6 and AOS8. Let
> me know if this is not correct.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ktbyers/netmiko/pull/486#issuecomment-307487281>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AGiMefudHifbVe6p9Ho16QqCgAb9HCPJks5sCaX_gaJpZM4NvbP3>
> .
>

ktbyers commented:
Yes, no worries at all. I committed it into the dev branch so we should be
good to go.



Kirk

On Fri, Jun 9, 2017 at 1:17 PM, gmcms <notifications@github.com> wrote:

> Hi Kirk, sorry I was not able to answer to you. I was in a DataCenter far
> from home without any access to my own environment.
>
> "Also can you add a note to the alcatel_aos_ssh.py file pertaining to which
> versions of AOS you tested with.
> From your earlier statements, I assumed this supported AOS6 and AOS8. Let
> me know if this is not correct."
>
> Yes it's ok for the AOS6 and AOS8. AOS7 is the same train as AOS8 which is
> the new release.
>
>
> "Does AOS do any paging of output...for example prompt for more?"
>
> No, the default is not to use paging. If you want to use it, you have a
> command syntax which is not the same on AOS6 and AOS8.
>
>
> "General pattern in Netmiko is not to emulate config mode (if it doesn't
> exist). So I think we should get rid of the self._config_mode variable and
> not track this.
> FYI, I removed the tracking of the config mode (since AOS doesn't support a
> config mode). Let me know if you see any problems with this."
>
> Sorry I thought that it was the right way to do this. I'm ok with your
> commit.
>
> If I can do more tests to add features or if you want to look at a way to
> add it in NAPALM, you can contact me.
>
>
> Thanks
>
>
> 2017-06-09 22:07 GMT+02:00 Kirk Byers <notifications@github.com>:
>
> > @gmcms <https://github.com/gmcms> FYI, I removed the tracking of the
> > config mode (since AOS doesn't support a config mode). Let me know if you
> > see any problems with this.
> >
> > See changes here ab3458a
> > <https://github.com/ktbyers/netmiko/commit/
> ab3458a64062038a41bb9d32e79fe3953684872f>
> >
> > From your earlier statements, I assumed this supported AOS6 and AOS8. Let
> > me know if this is not correct.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/ktbyers/netmiko/pull/486#issuecomment-307487281>,
> or mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/
> AGiMefudHifbVe6p9Ho16QqCgAb9HCPJks5sCaX_gaJpZM4NvbP3>
> > .
> >
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/ktbyers/netmiko/pull/486#issuecomment-307489467>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA26hCr1N1UiyoaVL2ur4T4mQiGHfyDyks5sCah0gaJpZM4NvbP3>
> .
>



-- 
Kirk Byers
CCIE #6243 emeritus
ktbyers@twb-tech.com
*http://pynet.twb-tech.com <http://pynet.twb-tech.com/blog>*
(415) 264-6270
Simplify your network management through automation

bacada105 commented:
Hello, when working with AOS 6 with Netmiko I found that older versions of AOS 6 like that on OS6400 and 6850 return a non standard host key lengths other then [1024, 2048, 3072] . Unfortunately the only workaround I found was the over ride below. This doesn't seam to be an issue on switches that run the latest versions of AOS like the 6450, 6250 etc... It would be worth it to note that these switches might have issues.

Ryan

`# from cryptography.hazmat.primitives.asymmetric import dsa
#
#
# def _override_check_dsa_parameters(parameters):
#     """Override check_dsa_parameters from cryptography's dsa.py
#
#     Allows for shorter or longer parameters.p to be returned from the server's host key. This is a
#     HORRIBLE hack and a security risk, please remove if possible!
#     """
#     # if utils.bit_length(parameters.p) not in [1024, 2048, 3072]:
#     # raise ValueError("p is {}, must be exactly 1024, 2048, or 3072 bits long".format(utils.bit_length(parameters.p)))
#     if dsa.utils.bit_length(parameters.q) not in [160, 256]:
#         raise ValueError("q must be exactly 160 or 256 bits long")
#
#     if not (1 < parameters.g < parameters.p):
#         raise ValueError("g, p don't satisfy 1 < g < p.")
#
# dsa._check_dsa_parameters = _override_check_dsa_parameters`

gmcms commented:
Hi OS6400 and OS6850 are no more supported by Alcatel-Lucent Enterprise.
Customers using those switches are pleased to move to new generations. It
would be better to add a comment rather than trying to modify netmiko.
Thanks
Regards

Sent from my custom Nexus

Le 17 juin 2017 12:03 AM, "bacada105" <notifications@github.com> a écrit :

> Hello, when working with AOS 6 with Netmiko I found that older versions of
> AOS 6 like that on OS6400 and 6850 return a non standard host key lengths
> other then [1024, 2048, 3072] . Unfortunately the only workaround I found
> was the over ride below. This doesn't seam to be an issue on switches that
> run the latest versions of AOS like the 6450, 6250 etc... It would be worth
> it to note that these switches might have issues.
>
> Ryan
>
> `# from cryptography.hazmat.primitives.asymmetric import dsa
> def _override_check_dsa_parameters(parameters): """Override
> check_dsa_parameters from cryptography's dsa.py Allows for shorter or
> longer parameters.p to be returned from the server's host key. This is a HORRIBLE
> hack and a security risk, please remove if possible! """ # if
> utils.bit_length(parameters.p) not in [1024, 2048, 3072]: # raise
> ValueError("p is {}, must be exactly 1024, 2048, or 3072 bits
> long".format(utils.bit_length(parameters.p))) if dsa.utils.bit_length(parameters.q)
> not in [160, 256]: raise ValueError("q must be exactly 160 or 256 bits
> long") if not (1 < parameters.g < parameters.p): raise ValueError("g, p
> don't satisfy 1 < g < p.") dsa._check_dsa_parameters =
> _override_check_dsa_parameters`
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ktbyers/netmiko/pull/486#issuecomment-309142453>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AGiMefnZNoFLIuKAEWyEKMKTW0BV7sG6ks5sEvivgaJpZM4NvbP3>
> .
>

ktbyers commented:
@bacada105 Does that sound reasonable to you? We will add a note in the Netmiko source code pertaining to this issue?

bacada105 commented:
I'd agree that a comment would be best as those models are end of life. This would avoid confusion   of supposed models running AOS 6.

LensHunnel commented:
There is a lot of bugs in AOS6 on 6450 and 6850E which are always supported by Alcatel - Lucent.
